components/merkledb/examples/migration_with_iter_loop.rs

```diff
@@ -0,0 +1,162 @@
+//! 4. We complete the migration by calling `flush_migration`. This moves the migrated data
+//!   to its intended place and removes the old data marked for removal.
+
+use exonum_merkledb::{
+    access::Prefixed,
+    migration::{flush_migration, MigrationHelper},
+    ObjectHash, Result as DbResult,
+};
+
+use migration::{
+    check_data_after_flush, check_data_after_merge, check_data_before_flush, create_initial_data,
+    v1, v2,
+};
+
+mod migration;
+
+fn create_migration(helper: &mut MigrationHelper) -> DbResult<()> {
+    println!("\nStarted migration");
+
+    {
+        let new_data = helper.new_data();
+        let old_data = helper.old_data();
+
+        let old_schema = v1::Schema::new(old_data);
+        let mut new_schema = v2::Schema::new(new_data);
+
+        // Move `ticker` and `divisibility` to `config`.
+        let config = v2::Config {
+            ticker: old_schema.ticker.get().unwrap(),
+            divisibility: old_schema.divisibility.get().unwrap_or(0),
+        };
+        new_schema.config.set(config);
+        // Mark these two indexes for removal.
+        new_data.create_tombstone("ticker");
+        new_data.create_tombstone("divisibility");
+    }
+
+    // Migrate wallets using `MigrationHelper::iter_loop`:
+    // `Wallet::public_key` field will be removed.
+    // `Wallet::history_hash` field will be added.
+    // Wallets and history from username Eve will be removed.
+    migrate_wallets(helper)
+}
+
+/// Provides migration of wallets with `MigrationHelper::iter_loop`.
+/// `iter_loop` is designed to allow to merge changes to the database from time to time,
+/// so we are migrating wallets in chunks here.
+fn migrate_wallets(helper: &mut MigrationHelper) -> DbResult<()> {
+    helper.iter_loop(|helper, iters| {
+        let old_schema = v1::Schema::new(helper.old_data());
+        let mut new_schema = v2::Schema::new(helper.new_data());
+
+        const CHUNK_SIZE: usize = 1_000;
+        for (public_key, wallet) in iters
+            .create("wallets", &old_schema.wallets)
+            .take(CHUNK_SIZE)
+        {
+            if wallet.username == "Eve" {
+                // We don't like Eves 'round these parts. Remove her transaction history
+                // and don't migrate the wallet.
+                helper
+                    .new_data()
+                    .create_tombstone(("histories", &public_key));
+            } else {
+                // Merkelize the wallet history.
+                let mut history = new_schema.histories.get(&public_key);
+                history.extend(&old_schema.histories.get(&public_key));
+
+                let new_wallet = v2::Wallet {
+                    username: wallet.username,
+                    balance: wallet.balance,
+                    history_hash: history.object_hash(),
+                };
+                new_schema.wallets.put(&public_key, new_wallet);
+            }
+        }
+
+        println!("Processed chunk of {} wallets", CHUNK_SIZE);
```

Comment:
Nit: This number isn't necessarily correct on the last iteration (e.g., we have 3456 wallets, then on the last iteration we'll process 456 of them).